---
title: "Differential Gene Expression Analysis - ADNI RNA"
author: "Christelle Schneuwly Diaz"
date: "2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

# Differential expression analysis 

In this report, we perform classical differential gene expression analysis on ADNI microarray data. After preprocessing and filtering, we compare diagnostic groups (e.g., CN vs. AD, CN vs. MCI) using edgeR. Visualizations include volcano plots to highlight significantly deregulated genes.

```{r Import packages and data}
library(ADNIMERGE)

library(MGMM)

library(edgeR)
library(locfit)
library(limma)
library(DESeq2)
library(mixtools)

library(dplyr)
library(tidyverse)
library(RColorBrewer)
library(pheatmap)
library(broom)
library(rootSolve)
library(Hmisc)

library(ggplot2)
library(ggrepel)
library(reshape2)
library(ggfortify)

library(samr)

filtered_gene_data <- read.csv("../data/microarray/filtered_counts.csv", 
                               row.names = 1, 
                               check.names = FALSE,
                               stringsAsFactors = FALSE)
filtered_gene_info <- read.csv("../data/microarray/filtered_genes.csv", 
                                row.names = 1, 
                               check.names = FALSE,
                                stringsAsFactors = FALSE)
filtered_samples <- read.csv("../data/microarray/filtered_samples.csv", 
                              row.names = 1, 
                             check.names = FALSE,
                              stringsAsFactors = FALSE)
df_dge_original <- read.csv("/home/chris/Documents/CHUV/PhD_project/dataset/ADNI/gene_expression_microarray/dge_gene_selection.csv", 
                                row.names = 1, 
                                check.names = FALSE,
                                stringsAsFactors = FALSE)
  
# Remove problematic samples
exclude_ids <- c("128_S_2002", "003_S_2374", "116_S_4167", "033_S_4176", 
                 "033_S_4179", "098_S_4215", "941_S_4292", "073_S_4300", 
                 "018_S_4349", "116_S_4453", "135_S_4489", "033_S_4505")

gene_counts <- filtered_gene_data[, !colnames(filtered_gene_data) %in% exclude_ids]
sample_meta <- filtered_samples[!filtered_samples$SubjectID %in% exclude_ids, ]
gene_info <- filtered_gene_info
rownames(gene_info) <- gene_info$Symbol
```

```{r FromEdgeR}
dge <- DGEList(counts = gene_counts, samples = sample_meta, genes = gene_info, group = factor(sample_meta$DX))
design <- model.matrix(~ 0 + factor(sample_meta$DX))

plot_volcano <- function(dge_result, threshold = 0.6, title = "Volcano Plot", highlight_genes = NULL) {
  dge_result$table$color_group <- ifelse(
    -log10(dge_result$table$PValue) > threshold & dge_result$table$logFC > 0, "Upregulated",
    ifelse(-log10(dge_result$table$PValue) > threshold & dge_result$table$logFC < 0, "Downregulated", "Not significant")
  )
  
  highlight_df <- dge_result$table[rownames(dge_result$table) %in% highlight_genes & -log10(dge_result$table$PValue) > threshold, ]

  ggplot(dge_result$table, aes(x = logFC, y = -log10(PValue))) +
    geom_point(aes(color = color_group), alpha = 0.8, size = 2.5) +
    geom_hline(yintercept = threshold, linetype = "dashed", color = "grey40") +
    scale_color_manual(values = c("Upregulated" = "red", "Downregulated" = "blue", "Not significant" = "grey")) +
    labs(x = "Log2 Fold Change", y = "-Log10 P-Value", title = title) +
    theme_minimal(base_size = 14) +
    theme(legend.position = "none") +
    geom_text_repel(
      data = highlight_df,
      aes(label = rownames(highlight_df)),
      size = 3.5,
      box.padding = 0.4,
      point.padding = 0.3,
      segment.color = 'grey50',
      max.overlaps = 15
    )
}


disp_cn_ad <- estimateDisp(dge, design)
et_cn_ad <- exactTest(disp_cn_ad, pair = c("CN", "AD"))
top_cn_ad <- topTags(et_cn_ad, n = Inf, sort.by = "none")

highlight_genes_cn_ad <- df_dge_original$Symbol[df_dge_original$Test == "CN_AD"]

plot_volcano(top_cn_ad, threshold = 0.6, title = "CN vs AD", highlight_genes = highlight_genes_cn_ad)

disp_cn_mci <- estimateDisp(dge, design)
et_cn_mci <- exactTest(disp_cn_mci, pair = c("CN", "MCI"))
top_cn_mci <- topTags(et_cn_mci, n = Inf, sort.by = "none")

highlight_genes_cn_mci <- df_dge_original$Symbol[df_dge_original$Test == "CN_MCI"]

plot_volcano(top_cn_mci, threshold = 0.6, title = "CN vs MCI", highlight_genes = highlight_genes_cn_mci)


disp_mci_ad <- estimateDisp(dge, design)
et_mci_ad <- exactTest(disp_mci_ad, pair = c("MCI", "AD"))
top_mci_ad <- topTags(et_mci_ad, n = Inf, sort.by = "none")

highlight_genes_mci_ad <- df_dge_original$Symbol[df_dge_original$Test == "CN_MCI"]

plot_volcano(top_mci_ad, threshold = 0.6, title = "CN vs MCI", highlight_genes = highlight_genes_mci_ad)
```

```{r Extracting top genes}
# Threshold
dge_threshold <- 0.6

# CN vs AD
sig_cn_ad <- subset(top_cn_ad$table, -log10(PValue) > dge_threshold)
top_genes_cn_ad <- rownames(sig_cn_ad)

# CN vs MCI
sig_cn_mci <- subset(top_cn_mci$table, -log10(PValue) > dge_threshold)
top_genes_cn_mci <- rownames(sig_cn_mci)

# MCI vs AD
sig_mci_ad <- subset(top_mci_ad$table, -log10(PValue) > dge_threshold)
top_genes_mci_ad <- rownames(sig_mci_ad)

plot_heatmap <- function(gene_list, title, dge_obj) {
  # Get expression matrix (logCPM)
  expr_matrix <- cpm(dge_obj, log = TRUE)[gene_list, ]

  # Z-score normalization (row-wise)
  z_expr <- t(scale(t(expr_matrix)))

  # Sample group annotations
  annotation_col <- data.frame(
    Diagnosis = dge_obj$samples$group
  )
  rownames(annotation_col) <- colnames(z_expr)

  # Plot heatmap
  pheatmap(
    z_expr,
    annotation_col = annotation_col,
    show_rownames = TRUE,
    show_colnames = FALSE,
    color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(100),
    cluster_cols = TRUE,
    cluster_rows = TRUE,
    fontsize_row = 7,
    main = title
  )
}

plot_heatmap(top_genes_cn_ad, title = "Top DE Genes: CN vs AD", dge_obj = dge)
plot_heatmap(top_genes_cn_mci, title = "Top DE Genes: CN vs MCI", dge_obj = dge)
plot_heatmap(top_genes_mci_ad, title = "Top DE Genes: MCI vs AD", dge_obj = dge)

sig_cn_ad$Test <- "CN_AD"
sig_mci_ad$Test <- "MCI_AD"
sig_cn_mci$Test <- "CN_MCI"

sig_cn_ad$Symbol <- rownames(sig_cn_ad)
sig_mci_ad$Symbol <- rownames(sig_mci_ad)
sig_cn_mci$Symbol <- rownames(sig_cn_mci)

df_dge <- rbind(sig_cn_ad, sig_mci_ad, sig_cn_mci)
df_dge <- data.frame(lapply(df_dge, as.character), stringsAsFactors=FALSE)

write.csv(df_dge, "../data/microarray/dge_gene_selection_edger.csv", row.names = TRUE)
```

```{r SAMR}

sam <- SAM(x=filtered_gene_data, y=as.factor(filtered_meta_data$DX), geneid = rownames(filtered_gene_info), genenames=filtered_gene_info$Symbol, resp.type="Multiclass", fdr.output=0.05)

# plot results
print(sam)
plot(sam)

if (median_filtering) {
write.csv(sam[["siggenes.table"]][["genes.up"]], "/home/chris/Documents/CHUV/PhD_project/dataset/ADNI/gene_expression_microarray/samr_gene_up.csv", row.names = TRUE)
}

# Function to run SAM analysis for a pairwise comparison
run_sam_pairwise <- function(gene_data, gene_info, meta_data, group1, group2) {
  # Subset the data for the two groups
  subset_indices <- which(meta_data$DX %in% c(group1, group2))
  subset_gene_data <- gene_data[, subset_indices]
  subset_meta_data <- meta_data[subset_indices, ]
  
  # Convert the response to numeric (1 for group1, 2 for group2)
  y <- as.numeric(as.factor(subset_meta_data$DX))
  
  # Prepare the SAM input data
  sam_data <- list(
    x = as.matrix(subset_gene_data),
    y = y,
    geneid = rownames(subset_gene_data),
    genenames = gene_info$Symbol,
    logged2 = TRUE  # Assuming your data is log-transformed
  )
  
  # Run SAM analysis
  samfit <- samr(
    sam_data,
    resp.type = "Two class unpaired",
    nperms = 100  # Number of permutations, adjust as needed
  )
  
  # Compute significant genes
  delta_table <- samr.compute.delta.table(samfit)
  significant_genes <- samr.compute.siggenes.table(samfit, del = 0.4, data = sam_data, delta.table = delta_table)
  
  # Plot the results
  samr.plot(samfit, 0.4)
  
  return(significant_genes)
}

#Run pairwise comparisons
# CN vs MCI
significant_genes_cn_mci <- run_sam_pairwise(filtered_gene_data, filtered_gene_info, filtered_meta_data, "CN", "MCI")

# CN vs AD
significant_genes_cn_ad <- run_sam_pairwise(filtered_gene_data, filtered_gene_info, filtered_meta_data, "CN", "AD")

# MCI vs AD
significant_genes_mci_ad <- run_sam_pairwise(filtered_gene_data, filtered_gene_info, filtered_meta_data, "MCI", "AD")

# You can access the significant genes for each comparison
sig_genes_cn_mci_up <- significant_genes_cn_mci$genes.up
sig_genes_cn_mci_down <- significant_genes_cn_mci$genes.lo

sig_genes_cn_ad_up <- significant_genes_cn_ad$genes.up
sig_genes_cn_ad_down <- significant_genes_cn_ad$genes.lo

sig_genes_mci_ad_up <- significant_genes_mci_ad$genes.up
sig_genes_mci_ad_down <- significant_genes_mci_ad$genes.lo

```